File: Sources/SomnioServerCore/Connections/CheckpointService.swift
import Foundation
import Logging
import ServiceLifecycle
import SomnioData

/// Periodic per-character checkpoint timer. Each pass calls `WorldRouter.checkpointAll()`; the
/// shutdown-cancel + error semantics live in `runPeriodically`.
///
/// The same pass sweeps expired session rows. They ride along here rather than in a service of their
/// own because the work is identical in shape — a periodic database write with no coordination — and
/// a second timer would double the machinery for one statement. The sweep is what makes
/// `SessionRepository.deleteExpired`'s "keeps the table from growing without bound" true: nothing on
/// the connection path can remove a row, because refresh, network loss, and a normal close must all
/// *preserve* the token for resumption.
public actor CheckpointService: Service {
    private let worldRouter: WorldRouter
    private let sessions: any SessionRepository
    private let interval: Duration
    private let logger: Logger

    public init(
        worldRouter: WorldRouter,
        sessions: any SessionRepository,
        interval: Duration,
        logger: Logger
    ) {
        self.worldRouter = worldRouter
        self.sessions = sessions
        self.interval = interval
        self.logger = logger
    }

    public func run() async throws {
        await runPeriodically(interval: interval, logger: logger, label: "checkpoint") { [worldRouter, sessions, logger] in
            await worldRouter.checkpointAll()
            // Failure is logged and swallowed, never propagated: an unreachable database must not
            // take the checkpoint timer down with it, and an unswept row is harmless until the next
            // pass. Expiry is enforced on the read path regardless, so a missed sweep costs storage
            // rather than letting a dead token authenticate.
            do {
                let deleted = try await sessions.deleteExpired(asOf: Date())
                if deleted > 0 {
                    logger.info("expired sessions swept", metadata: ["count": "\(deleted)"])
                }
            } catch {
                logger.warning("expired session sweep failed", metadata: ["error": "\(error)"])
            }
        }
    }
}

File: packages/server/src/services/checkpointService.ts
import type { SessionRepository } from '@somnio/data';
import type { Logger } from '../logging.ts';
import type { WorldRouter } from '../world/worldRouter.ts';
import { runPeriodically } from './periodicLoop.ts';

/**
 * The periodic checkpoint plus the expired-session sweep, which rides along because it is the
 * same shape of work and nothing on the connection path removes an expired session row.
 */
export class CheckpointService {
  private readonly worldRouter: WorldRouter;
  private readonly sessions: SessionRepository;
  private readonly intervalMs: number;
  private readonly logger: Logger;

  constructor(worldRouter: WorldRouter, sessions: SessionRepository, intervalMs: number, logger: Logger) {
    this.worldRouter = worldRouter;
    this.sessions = sessions;
    this.intervalMs = intervalMs;
    this.logger = logger;
  }

  run(signal: AbortSignal): Promise<void> {
    return runPeriodically(this.intervalMs, signal, async () => {
      await this.worldRouter.checkpointAll();
      // Logged and swallowed: an unreachable database must not take the timer down, and an
      // unswept row is harmless until the next pass because expiry is enforced on read.
      try {
        const deleted = await this.sessions.deleteExpired(new Date());
        if (deleted > 0) this.logger.info({ count: deleted }, 'expired sessions swept');
      } catch (error) {
        this.logger.warn({ error: String(error) }, 'expired session sweep failed');
      }
    });
  }
}
